Compactor: Document reasons and solutions about overlaps (#2191)
thanos-io/thanos: docs/operating/troubleshooting.md

slug: /troubleshooting.md
---

> # Overlaps

Reviewer: ```suggestion
# Troubleshooting; Common cases

## Compactor: Overlapping blocks

```

> Overlaps should never happen in healthy and well configured Thanos system. That's why there is no automatic repair.

Reviewer: Can we give example log line you see when you have overlaps? We should document each field in this log line as it can be quite tricky! (: WDYT?

Author: That would be helpful!
But i do not see any overlaps when use compator :( 
I would use logs in related issues.

Author: I only find those logs, could you provide more if possible?

Reviewer: Just one is enough. But would be nice to expand it and explain. 

```msg="critical error detected; halting" err="compaction failed: compaction: pre compaction overlap check: overlaps found while gathering blocks. [mint: 1555128000000, maxt: 1555135200000, range: 2h0m0s, blocks: 2]: <ulid: 01D94ZRM050JQK6NDYNVBNR6WQ, mint: 1555128000000, maxt: 1555135200000, range: 2h0m0s>, <ulid: 01D8AQXTF2X914S419TYTD4P5B, mint: 1555128000000, maxt: 1555135200000, range: 2h0m0s>```

This example shows that you have really the same blocks for 2h overlaps. It worth checking are those blocks the same (check the number of series and samples, checksum the index and chunks).

Another example would be:
```
sg="critical error detected; halting" err="compaction failed: compaction: pre compaction overlap check: overlaps found while gathering blocks. [mint: 1555128000000, maxt: 1555135200000, range: 2h0m0s, blocks: 2]: <ulid: 01D94ZRM050JQK6NDYNVBNR6WQ, mint: 1555128000000, maxt: 1555135200000, range: 2h0m0s>, <ulid: 01D8AQXTF2X914S419TYTD4P5B, mint: 1555128000000, maxt: 1555142400000, range: 4h0m0s>
```

In such example we can explain that there are two blocks overlapping in 2h time range. One 4h second 2h block (: It's more tricky. This suggests someone uploaded compacted blocks rather. Again looking on meta.json would help.

slug: /troubleshooting.md
---

# Troubleshooting; Common cases

## Overlaps

When overlaps happened, you can see logs like:
- "overlaps found while gathering blocks. [mint: 1582567200000, maxt: 1582574400000, range: 2h0m0s, blocks: 2]: <ulid> ..."

> - "overlaps found dir. /var/lib/prometheus/01E23JK5BP06RF1X519QS0JMXF/chunks"

Reviewer: This is not the correct log I think.

slug: /troubleshooting.md
---

# Troubleshooting; Common cases

## Overlaps

> Overlaps should never happen in healthy and well configured Thanos system. That's why there is no automatic repair.

Reviewer: ```suggestion
**Block overlap**: Set of blocks with exactly the same external labels in meta.json and for the same time or overlapping time period.

Thanos is designed to never end up with overlapped blocks. This means that (uncontrolled) block overlap should never happen in a healthy and well configured Thanos system. That's why there is no automatic repair for this. Since it's an unexpected incident:
* All reader components like Store Gateway will handle this gracefully (overlapped samples will be deduplicated).
* Thanos compactor will stop all activities and HALT or crash (with metric and will error log). This is because it cannot perform compactions and downsampling. In the overlap situation, we know something unexpected happened (e.g manual block upload, some malformed data etc), so it's safer to stop or crash loop (it's configurable).
```

Let's take two examples:

- `msg="critical error detected; halting" err="compaction failed: compaction: pre compaction overlap check: overlaps found while gathering blocks. [mint: 1555128000000, maxt: 1555135200000, range: 2h0m0s, blocks: 2]: <ulid: 01D94ZRM050JQK6NDYNVBNR6WQ, mint: 1555128000000, maxt: 1555135200000, range: 2h0m0s>, <ulid: 01D8AQXTF2X914S419TYTD4P5B, mint: 1555128000000, maxt: 1555135200000, range: 2h0m0s>`

> It shows that you have really the same blocks for 2h overlaps. It worth checking whether are those blocks the same (check the number of series and samples, checksum the index and chunks).

Reviewer: ```suggestion
In this halted example, we can read that compactor detected 2 overlapped blocks. What's interesting is that those two blocks look like they are "similar". They are exactly for the same period of time. This might mean that potential reasons are:
* Duplicated upload with different ULID (non-persistent storage for Prometheus can cause this)
* 2 Prometheus-es are misconfigured and they are uploading the data with exactly the same external labels. This is wrong, they should be unique.

Checking producers log for such ULID, and checking meta.json (e.g if sample stats are the same or not) helps.  Checksum the index and chunks files as well to reveal if data is exactly the same, thus ok to be removed manually.
```

- `msg="critical error detected; halting" err="compaction failed: compaction: pre compaction overlap check: overlaps found while gathering blocks. [mint: 1555128000000, maxt: 1555135200000, range: 2h0m0s, blocks: 2]: <ulid: 01D94ZRM050JQK6NDYNVBNR6WQ, mint: 1555128000000, maxt: 1555135200000, range: 2h0m0s>, <ulid: 01D8AQXTF2X914S419TYTD4P5B, mint: 1555128000000, maxt: 1555142400000, range: 4h0m0s>`

> It shows there are two blocks overlapping in 2h time range. One is 4h, the other is 2h.It's more tricky. This suggests you uploaded compacted blocks rather. Again looking on meta.json would help.

Reviewer: ```suggestion
This case is more tricky. It shows there are two blocks overlapping in 2h time range. One is 4h, the other is 2h. This might suggest a manual uploaded of compacted blocks or multiple compactors (which is wrong). Again looking on meta.json would help.
```

> You can read following reasons and choose a right solution to fix it by hand.

Reviewer: Let's remove this - we will never create a list of all possible problems.
